core/src/main/java/org/apache/iceberg/avro/TypeToSchema.java

```diff
@@ -112,7 +112,9 @@ public Schema struct(Types.StructType struct, List<Schema> fieldSchemas) {
     }
 
     recordSchema = Schema.createRecord(recordName, null, null, false, fields);
-
+    if (struct.isUnionSchema()) {
```

Comment:
Is there some reason why this needs to be a round-trip conversion? I think it should be fine to convert back to a record instead of a union schema. That's what we should be writing into a table with a schema converted from an Avro union schema anyway.

Reply:
this is needed when avro writer is java client which is writing directly by consuming avro records.

Reply:
for spark , spark avro convert its into record so no changes needed there. I will add test for spark

Reply:
only way for `GenericAvroWriter` to know if record represents union schema type is to use property set here. otherwise ValueWriters will fail to write union schema record as they will expect new record schema instead of union record schema.